_posts/2021-05-06-broker-tuning.md

```diff
@@ -0,0 +1,443 @@
+
+### Improving request handling throughput by increasing I/O threads
+
+Network threads (`num.network.threads`) handle requests to the Kafka cluster, such as produce and fetch requests from client applications.
+Adjust the number of network threads to reflect the replication factor and the levels of activity from client producers and consumers interacting with the Kafka cluster.
+
+I/O threads (`num.io.threads`) pick up requests from the request queue to process them. Adding more threads can improve throughput, but the number of CPU cores/and disk bandwidth imposes a practical upper limit.
+The minimum number of threads should equal the number of storage volumes.
+
+To reduce congestion and regulate the request traffic, you can use `queued.max.requests` to limit the number of requests allowed in the request queue before the network thread is blocked.
+Use `num.recovery.threads.per.data.dir` to specify the number of threads used for log loading at startup and flushing at shutdown.
+
+```properties
+# ...
+num.io.threads=8
+queued.max.requests=500
+num.network.threads=3
+num.recovery.threads.per.data.dir=1
+# ...
+```
+
+> Kafka broker metrics can help with working out the number of threads required. For example, metrics for the average time network threads are idle (`kafka.network:type=SocketServer,name=NetworkProcessorAvgIdlePercent`) indicate the percentage of resources used.
+If there is 0% idle time, all resources are in use, which means that more threads would be beneficial.
+
+If threads are slow or limited due to the number of disks, try increasing the size of the buffers for network requests to improve throughput:
+
+```properties
+# ...
+replica.socket.receive.buffer.bytes=65536
+# ...
+```
+
+And also increase the maximum number of bytes Kafka can receive:
+
+```properties
+# ...
+socket.request.max.bytes=104857600
+# ...
+```
+
+### Increasing bandwidth for high latency connections
+
+If you've fine-tuned the size of your message batches, the default values of the buffers for sending and receiving messages might be too small for the required throughput.
+
+```properties
+# ...
+socket.send.buffer.bytes=1048576
+socket.receive.buffer.bytes=1048576
+# ...
+```
+
+You can work out the optimal size of your buffers using a _bandwidth-delay_ product calculation.
+You need some figures to work with. Essentially, you want to multiply your bandwidth capacity with the round-trip delay to give the data volume.
+Go look at this [wiki definition](https://en.wikipedia.org/wiki/Bandwidth-delay_product) to see examples of the calculation.
+
+### Managing logs with data retention policies
+
+Kafka uses logs to store message data. Logs are a series of segments. New messages are written to a single _active_ segment. The messages are never modified.  
+Using fetch requests, consumers read from the active segment. After a time, the active segment is _rolled_ to become ready-only as a new active segment replaces it.
```

Comment:
Consumers don't always ready from the active segment, they can read from any.

Reply:
Changed to:

> Using fetch requests, consumers read from the segments.